File: opencv-sandbox/homography.cpp
#include <stdio.h>
#include <iostream>
#include "opencv2/core.hpp"
#include "opencv2/imgproc.hpp"
#include "opencv2/features2d.hpp"
#include "opencv2/highgui.hpp"
#include "opencv2/calib3d.hpp"
#include "opencv2/xfeatures2d.hpp"
using namespace cv;
using namespace cv::xfeatures2d;
void readme();
/* @function main */
int main( int argc, char** argv )
{
  // READ FILES
  if( !(argc == 3 || argc == 4 ))
  { readme(); return -1; }
  Mat img_object = imread( argv[1]);
  Mat img_scene = imread( argv[2]);
  if( !img_object.data || !img_scene.data )
  { std::cout<< " --(!) Error reading images " << std::endl; return -1; }

  // CREATE DETECTOR
  Ptr<SIFT> detector = SIFT::create();
  std::vector<KeyPoint> keypoints_object, keypoints_scene;
  Mat descriptors_object, descriptors_scene;

  // DETECT KEY POINTS AND DESCRIPTORS
  detector->detectAndCompute( img_object, Mat(), keypoints_object, descriptors_object );
  detector->detectAndCompute( img_scene, Mat(), keypoints_scene, descriptors_scene );

  if ( descriptors_object.empty() ) {
     cvError(0,"MatchFinder","object descriptor empty",__FILE__,__LINE__);
  }
  if ( descriptors_scene.empty() ) {
     cvError(0,"MatchFinder","scene descriptor empty",__FILE__,__LINE__);
  }

  // MATCH
  FlannBasedMatcher matcher;
  std::vector< DMatch > matches;
  matcher.match( descriptors_object, descriptors_scene, matches );
  double max_dist = 0; double min_dist = 100;

  for( int i = 0; i < descriptors_object.rows; i++ )
  { double dist = matches[i].distance;
    if( dist < min_dist ) min_dist = dist;
    if( dist > max_dist ) max_dist = dist;
  }
  printf("-- Max dist : %f \n", max_dist );
  printf("-- Min dist : %f \n", min_dist );
  printf("-- Match count : %i \n", descriptors_object.rows );

  // DRAW MATCHES
  Mat img_matches;
  drawMatches( img_object, keypoints_object, img_scene, keypoints_scene,
               matches, img_matches, Scalar::all(-1), Scalar::all(-1),
               std::vector<char>());

  std::vector<Point2f> obj;
  std::vector<Point2f> scene;
  for( size_t i = 0; i < matches.size(); i++ )
  {
    obj.push_back( keypoints_object[ matches[i].queryIdx ].pt );
    scene.push_back( keypoints_scene[ matches[i].trainIdx ].pt );
  }

  Mat H = findHomography( obj, scene, RANSAC );
  if ( H.empty() ) {
     cvError(0,"MatchFinder","no no homography",__FILE__,__LINE__);
  }

  std::vector<Point2f> obj_corners(4);
  obj_corners[0] = cvPoint(0,0);
  obj_corners[1] = cvPoint( img_object.cols, 0 );
  obj_corners[2] = cvPoint( img_object.cols, img_object.rows );
  obj_corners[3] = cvPoint( 0, img_object.rows );
  std::vector<Point2f> scene_corners(4);
  perspectiveTransform( obj_corners, scene_corners, H);

  line( img_matches, scene_corners[0] + Point2f( img_object.cols, 0), scene_corners[1] + Point2f( img_object.cols, 0), Scalar(0, 255, 0), 4 );
  line( img_matches, scene_corners[1] + Point2f( img_object.cols, 0), scene_corners[2] + Point2f( img_object.cols, 0), Scalar( 0, 255, 0), 4 );
  line( img_matches, scene_corners[2] + Point2f( img_object.cols, 0), scene_corners[3] + Point2f( img_object.cols, 0), Scalar( 0, 255, 0), 4 );
  line( img_matches, scene_corners[3] + Point2f( img_object.cols, 0), scene_corners[0] + Point2f( img_object.cols, 0), Scalar( 0, 255, 0), 4 );

  if(argc == 3 ) {
    imshow( "Good Matches & Object detection", img_matches );
    waitKey(0);
  } else {
    imwrite(argv[3], img_matches);
  }
  return 0;
}

void readme()
{ std::cout << " Usage: ./SIFT_descriptor <img1> <img2>" << std::endl; }

File: Homography.hs
{-# LANGUAGE DataKinds #-}

module Main where
import Control.Monad
import Data.Default
import Data.Foldable
import Data.Int
import Data.Maybe
import qualified Data.Vector as V
import Data.Word
import Foreign.C.Types
import GHC.Float
import Linear
import OpenCV
import OpenCV.Extra.XFeatures2d
import OpenCV.Internal.C.Types

import System.IO.Unsafe ( unsafePerformIO )
import qualified Data.ByteString as B

import System.Environment

type DynMat = Mat ('S ['D, 'D]) ('S 3) ('S Word8)


main = do
  -- READ FILES
  let imread = (imdecode ImreadUnchanged <$>) . B.readFile
  [imgObject, imgScene] <- (take 2 <$> getArgs) >>= mapM imread

  -- CREATE DETECTOR
  let sift = mkSift defaultSiftParams

  -- DETECT KEY POINTS AND DESCRIPTORS
  let (keypointsObject, descriptorsObject) = exceptError $ siftDetectAndCompute sift imgObject Nothing
  let (keypointsScene, descriptorsScene) = exceptError $ siftDetectAndCompute sift imgScene Nothing

  -- MATCH
  fbmatcher <- newFlannBasedMatcher (def { indexParams = FlannKDTreeIndexParams 1 })
  matches <- match fbmatcher descriptorsObject descriptorsScene Nothing

  let showDist f = show $ f $ V.map (dmatchDistance . dmatchAsRec) matches
  putStrLn $ "-- Max dist : " ++ showDist maximum
  putStrLn $ "-- Min dist : " ++ showDist minimum
  putStrLn $ "-- Match count : " ++ (show $ V.length matches)

  -- DRAW MATCHES
  let imgMatches = exceptError $ drawMatches imgObject keypointsObject imgScene keypointsScene matches def

  -- FIND HOMOGRAPHY

  let (framePts, startPts) =  V.unzip $ V.map (getMatchingPoints keypointsObject keypointsScene) matches
  let (homography, _) = fromMaybe undefined $ exceptError $ findHomography framePts startPts (def { fhpMethod = FindHomographyMethod_RANSAC })

  let [w, h] = fmap fromIntegral . miShape . matInfo $ imgObject
  let objCorners = V.fromList [ V2 0 0, V2 h 0, V2 h w, V2 0 w]

  let sceneCorners = V.map ((round <$>) . fromPoint . toPoint) $ perspectiveTransform objCorners homography :: V.Vector (V2 Int32)

  -- DRAW HOMOGRAPHY
  let green = toScalar (V4 0 255 0 255 :: V4 Double)
  imgMatchesM <- thaw imgMatches

  let offsetCorner n = (sceneCorners V.! n) + (V2 (round h) 0) -- cheating because we know image is rotated
  let line' n imgM = line imgM (offsetCorner n) (offsetCorner ((n+1) `mod` 4)) green 4 LineType_AA 0
  line' 0 imgMatchesM
  line' 1 imgMatchesM
  line' 2 imgMatchesM
  line' 3 imgMatchesM
  imgMatches <- freeze imgMatchesM

  _:_:f <- getArgs
  case f of
    [] -> do
      withWindow  "Good Matches & Object detection (Haskell)" $ \window -> do
        imshow window imgMatches
        void $ waitKey 0
    [f] -> do
      let bs = exceptError $ imencode (OutputPng defaultPngParams) imgMatches
      B.writeFile f bs

getMatchingPoints :: V.Vector KeyPoint -> V.Vector KeyPoint -> DMatch -> (V2 CDouble, V2 CDouble)
getMatchingPoints keypointsObject keypointsScene dmatch =
  let matchRec = dmatchAsRec dmatch
      queryPt = keypointsObject V.! fromIntegral (dmatchQueryIdx matchRec)
      trainPt = keypointsScene V.! fromIntegral (dmatchTrainIdx matchRec)
      queryPtRec = keyPointAsRec queryPt
      trainPtRec = keyPointAsRec trainPt
     in (v2ToDouble $ kptPoint queryPtRec, v2ToDouble $ kptPoint trainPtRec)
  where v2ToDouble = fmap (toCDouble . float2Double)

